### DISCUSSION TITLE
Grain reconstruction for Large Area Map

### DISCUSSION THREAD

**Discussion by huseyinhzl** [mtex-toolbox/mtex#1809] [Ask Anything]
Hi All,

Recently, I have done a large area map to reconstruct parent beta grains. I have 20 different tiles making about 1x1mm area in ebsd. When I created ctf file, I came across about 6 million data points which make my analysis almost impossible. I am wondering if there is a way to do the analysis in a more efficient way. If so, what can I do? 

Thank you so much for any help

**Comment by kilir** [comment]
Hi,
why do you mean it is impossible? Pure computation time? Or do you get any errors somewhere?
Cheers,
Rüdiger

**Reply by huseyinhzl** [reply]
Hi Rudiger,

I did not get any error but it is about pure computation time. Basically, grain computation is running for almost 2 and is still not finished. 

_P.S. I am using a workstation with Xeon CPU and 64 GB of RAM_

**Reply by nyyssont** [reply]
Do you mean that the initial computation of grains, using the calcGrains command, takes a long time?

This is a known issue for maps with very large amounts of grains. There is a fix which will be introduced to main MTEX at some point. Check https://github.com/mtex-toolbox/mtex/pull/1808

**Reply by kilir** [reply]
For a large number of grains, this might take some time currently and multicore machines would not help (actually it might be slower since the step taking the longest time is single threaded and often processors with a massive amount of cores might run at a lower frequnency).

Some thing that does not cut down computation time, but might work for you as well if you really want to analyse these large datasets:
-  do all cleanup / grain computation etc. on a small subset of your map e.g. using `ebsd = ebsd(inpolygon(ebsd,[Xs Ys dX dY]))`
- next, save the reuslts your will need at any later point e.g. ebsd dataset with grainIDs and the corresponding grains using  save('filename.mat','ebsd','grains')
-  eventually set up a loop reading input files e.g. ctf, generate reasonable filenames and save the results as above
e.g.
```
% assuming you have your input in a directory called input
infiles = dir('input/*.ctf');

for i=1:length(infiles)

    ebsd= loadEBSD_ctf([infiles(i).folder '/' infiles(i).name],'additional options ...' ); % adapt properly
    % followed by whatever it takes to properly correct the coordinate
    % systems etc...

    % do something e.g. calc grains

    % save the results
    save([infiles(i).folder '/' infiles(i).name(1:end-7) '_grains_ebsd.mat'],'grains','ebsd')

end   
```
-  if it's just about grain computation for the moment, you can run many instances in parallel on the same machine if cores permit
- make sure you do not use the 'qHull' option. While it's faster, it's a very ancient version build into matlab which does not cope with large number of edges.
- once everything works, comment the line restricting everything to a subset and let it run till it's done while you enjoy your weekend in the meantime or think about something else

Cheers,
Rüdiger

**Reply by huseyinhzl** [reply]
> Do you mean that the initial computation of grains, using the calcGrains command, takes a long time?
> 
> This is a known issue for maps with very large amounts of grains. There is a fix which will be introduced to main MTEX at some point. Check #1808

calcGrains command took about 4 hours for 5457689 (about 5.5 million) data. I have 5 other large area maps with the same amount of data. With this calculating time, it will take a week I guess. I'll check #1808 and will run it again. Also, after the calcGrains command, I tried to run the parent grain reconstruction using Variant Graph-based algorithm, yet I encountered an error "out of memory" so I could not run the necessary iterations. So this is a new problem for me :)

**Reply by huseyinhzl** [reply]
> For a large number of grains, this might take some time currently and multicore machines would not help (actually it might be slower since the step taking the longest time is single threaded and often processors with a massive amount of cores might run at a lower frequnency).
> 
> Some thing that does not cut down computation time, but might work for you as well if you really want to analyse these large datasets:
> 
> * do all cleanup / grain computation etc. on a small subset of your map e.g. using `ebsd = ebsd(inpolygon(ebsd,[Xs Ys dX dY]))`
> * next, save the reuslts your will need at any later point e.g. ebsd dataset with grainIDs and the corresponding grains using  save('filename.mat','ebsd','grains')
> * eventually set up a loop reading input files e.g. ctf, generate reasonable filenames and save the results as above
>   e.g.
> 
> ```
> % assuming you have your input in a directory called input
> infiles = dir('input/*.ctf');
> 
> for i=1:length(infiles)
> 
>     ebsd= loadEBSD_ctf([infiles(i).folder '/' infiles(i).name],'additional options ...' ); % adapt properly
>     % followed by whatever it takes to properly correct the coordinate
>     % systems etc...
> 
>     % do something e.g. calc grains
> 
>     % save the results
>     save([infiles(i).folder '/' infiles(i).name(1:end-7) '_grains_ebsd.mat'],'grains','ebsd')
> 
> end   
> ```
> 
> * if it's just about grain computation for the moment, you can run many instances in parallel on the same machine if cores permit
> * make sure you do not use the 'qHull' option. While it's faster, it's a very ancient version build into matlab which does not cope with large number of edges.
> * once everything works, comment the line restricting everything to a subset and let it run till it's done while you enjoy your weekend in the meantime or think about something else
> 
> Cheers, Rüdiger

Hi Rudiger,

This was what I was thinking about. I will divide my large area into subsets so that I will do all the necessary steps on the subsets, and then i will stitch the resultant plots to get one big plot.

**Reply by kilir** [reply]
Also, in case your data permits, using `'unitCell'` as an option to ´calcGrains´ may be significantly faster.

**Reply by huseyinhzl** [reply]
> Also, in case your data permits, using `'unitCell'` as an option to ´calcGrains´ may be significantly faster.

I have been looking for it, but I could not understand the gridfy command. Is it possible to explain to me how to use gridfy together with 'unitCell'.

**Reply by kilir** [reply]
It's rather unrelated, you can use calcGrains on gridified or a normal list of ebsd data.

**Reply by huseyinhzl** [reply]
Thank you for your reply. I will try to divide my large area and then do the all the calculations
